cpp/include/cudf/io/functions.hpp

```diff
@@ -616,28 +766,140 @@ struct write_parquet_args {
   /// Set of columns to output
   table_view table;
   /// Optional associated metadata
-  const table_metadata* metadata;
+  const table_metadata* metadata = nullptr;
   /// Optionally return the raw parquet file metadata output
   bool return_filemetadata = false;
   /// Column chunks file path to be set in the raw output metadata
   std::string metadata_out_file_path;
 
+  explicit write_parquet_args(sink_info const& sink_, table_view const& table_)
+    : sink(sink_), table(table_)
+  {
+  }
+
+  friend class write_parquet_args_builder;
+
+ public:
   write_parquet_args() = default;
 
-  explicit write_parquet_args(sink_info const& sink_,
-                              table_view const& table_,
-                              const table_metadata* metadata_ = nullptr,
-                              compression_type compression_   = compression_type::AUTO,
-                              statistics_freq stats_lvl_ = statistics_freq::STATISTICS_ROWGROUP)
-    : sink(sink_),
-      compression(compression_),
-      stats_level(stats_lvl_),
-      table(table_),
-      metadata(metadata_)
+  /**
+   * @brief Build write_parquet_args.
+   *
+   * @param sink_ sink to use for writer output
+   * @param table_ Table to be written to output
+   *
+   * @return write_parquet_args_builder parquet args builder with all arguments
+   */
+  static write_parquet_args_builder build(sink_info const& sink, table_view const& table);
+
+  /**
+   * @brief Returns sink info
+   */
+  sink_info get_sink_info() const { return sink; }
+
+  /**
+   * @brief Returns compression type
+   */
+  compression_type get_compression_type() const { return compression; }
+
+  /**
+   * @brief Returns stats level
+   */
+  statistics_freq get_stats_level() const { return stats_level; }
+
+  /**
+   * @brief Returns table_view
+   */
+  table_view get_table() const { return table; }
+
+  /**
+   * @brief Returns metadata
+   */
+  const table_metadata* get_metadata() const { return metadata; }
+
+  /**
+   * @brief Returns True/False for filemetadata is requried or not
+   */
+  bool is_filemetadata_required() const { return return_filemetadata; }
```

Comment:
:+1: 
required or requested, but this is much better naming than some in the reader.

Reply:
I think all bool parameters should go through a single interface.

```
enum class toggle : uint8_t {
  FILEMETADATA_REQUIRED,
  HUGS_REQUIRED,
  USE_INFINITE_COMPRESSION,
  VOTE_FOR_PEDRO
};

bool get(toggle t);
void set(toggle t, bool enabled)

// alternatively, or additionally:
void enable(toggle t)
void disable(toggle t)
```

Then you just keep an array of bools, or a map, or whatever.

Reply:
@jrhemstad @codereport , What is opinion regarding usage of enums with gets and sets.

Reply:
I'm not quite following what the suggestion is. Can someone summarize with the two options?

Reply:
Use enums for set of members which have same type and use single API to set/get them.

Reply:
```
enum class boolean_param_id : int8_t {
    STRINGS_TO_CATEGORICAL, // Whether to store string data as categorical type
    USE_PANDAS_METADATA, // Whether to use PANDAS metadata to load columns
  };

  /**
   * @brief enum class for read_parquet_args size_type parameters
   */
  enum class size_type_param_id : int8_t {
    SKIP_ROWS, // Rows to skip from the start
    NUM_ROWS // Rows to read
  };

bool get (boolean_param_id param_id) const
  {
    switch (param_id){
      case boolean_param_id::STRINGS_TO_CATEGORICAL:
          return _strings_to_categorical;
      case boolean_param_id::USE_PANDAS_METADATA:
          return _use_pandas_metadata;
      default: CUDF_FAIL("Unsupported boolean_param_id enum");
    }
  }

read_parquet_args_builder& set(read_parquet_args::boolean_param_id param_id, bool val)
  {
    switch (param_id) {
        case read_parquet_args::boolean_param_id::STRINGS_TO_CATEGORICAL:
          args._strings_to_categorical = val;
          break;
        case read_parquet_args::boolean_param_id::USE_PANDAS_METADATA:
          args._use_pandas_metadata = val;
          break;
      default: CUDF_FAIL("Unsupported boolean_param_id enum");
    }

    return *this;
  }

```

Reply:
I'm initially not a fan of using a single `get/set` interface with enums to control what you are accessing/mutating. Perhaps if there were many more such attributes, but it seems like there's only 2-3 of them. It reads easier to me to have functions for each attribute.

Reply:
only csv has comparatively bigger bunch of options with with same type, but reset of the APIs have only couple of them.

Reply:
The problem that led to this is we wanted to avoid the linguistic difficulties of functions with names like `get_use_pandas_metadata()` and `get_strings_to_categorical`.  So I suggested differentiating the verb (`get`) from the thing being "verbed" (the parameter). 

I am not tied to this suggestion, but I would like to aim to avoid the linguistic confusion in whatever we end up with.